### PR TITLE
Sensor endpoints and multitier refacor

### PR DESCRIPTION
**Pull request by justinf34**
What was implemented
- CRUD endpoints for Sensor entity. Still not done since it Thing entity still needs to be implemented
- Refactored to follow multitier design pattern. It now follows repository, service and model structure. Also, uses interface to hide implementation of repositories and services. For now, only the Sensor endpoints follow the new design pattern but I will work with other members to refactor the endpoints they implemented.
- Implemented a better error handling

To run, will need to update .env which was sent to the Softies channel.

### CODE REVIEW

File: app/handlers/sensor.handler.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/gin-gonic/gin"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type SensorHandler struct {
+	snsr sensorSrv.SensorServiceInterface
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: snsr? Could rename this to something better.

File: app/handlers/sensor.handler.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func NewSensorAPI(sensorService sensorSrv.SensorServiceInterface) *SensorHandler {
+	return &SensorHandler{
+		snsr: sensorService,
+	}
+}
+
+func (h *SensorHandler) Create(c *gin.Context) {
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: use a better variable name than "h".

File: app/handlers/sensor.handler.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	// Check if sensor already exists
+	if h.snsr.IsSensorAlreadyExists(c.Request.Context(), newSensor.ThingID, newSensor.SID) {
+		result = utils.NewHTTPError(utils.SensorAlreadyExists)
+		utils.Response(c, http.StatusBadRequest, result)
+		return
+	}
+
+	// TODO Check if Thing exists
+
+	// Generate last updated
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: Do last update and generate ID together, no need for comments, the code is clear.

File: app/handlers/sensor.handler.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	newSensor.LastUpdate = utils.CurrentTimeInMilli()
+
+	// Generate id
+	newSensor.ID = bson.NewObjectId()
+
+	err := h.snsr.Create(c.Request.Context(), &newSensor)
+	var status int
+	if err != nil {
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: If would personally handle the success condition first. We are most interested in a success, error handling first is confusing. Just a semantics thing. When I read code, I care most about the core logic first, THEN how errors are handled.

File: app/handlers/sensor.handler.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	result := make(map[string]interface{})
+
+	if err != nil {
+		result = utils.NewHTTPError(utils.SensorsNotFound)
+		utils.Response(c, http.StatusBadRequest, result)
+	} else {
+		result = utils.SuccessPayload(sensors, "Succesfully retrieved sensors")
+		utils.Response(c, http.StatusOK, result)
+	}
+
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: Inconsistent spacing. I personally prefer the code to be as tight as possible.

File: app/handlers/sensor.handler.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func (h *SensorHandler) FindById(c *gin.Context) {
+
+	id := c.Param("id")
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: What id is this? I recommend naming to sensorId.

File: app/handlers/sensor.handler.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+}
+
+func (h *SensorHandler) FindByThingIdAndLastUpdate(c *gin.Context) {
+
+	result := make(map[string]interface{})
+
+	thingId := c.Param("thingId")
+	lastUpdateStr := c.Param("lastUpdate")
+	lastUpdate, err := strconv.ParseInt(lastUpdateStr, 10, 64)
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: What does 10, 64 do?

File: app/handlers/sensor.handler.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	}
+
+}
+
+func (h *SensorHandler) FindByThingIdAndLastUpdate(c *gin.Context) {
+
+	result := make(map[string]interface{})
+
+	thingId := c.Param("thingId")
+	lastUpdateStr := c.Param("lastUpdate")
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: Why not just pass the call directly into the next line over making an extra variable? You are using extra memory when you don't need to. Not sure if Go works like cpp where parameters are copied rather than passed by reference. You can pass the lastUpdateStr as an rvalue, you don't use it again.

File: app/handlers/sensor.handler.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+		result = utils.NewHTTPError(utils.SensorNotFound)
+		utils.Response(c, http.StatusBadRequest, result)
+	} else {
+		result = utils.SuccessPayload(sensor, "Succesfully retrieved sensor")
+		utils.Response(c, http.StatusOK, result)
+	}
+
+}
+
+func (h *SensorHandler) FindByThingIdAndLastUpdate(c *gin.Context) {
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: Why do we even need this endpoint? The hardware will send the most recent update of it's most recently updated sensor. Find all sensors that have a last update greater than the lastUpdate passed and return the list.

File: app/handlers/sensor.handler.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		utils.Response(c, http.StatusBadRequest, result)
+	} else {
+		result = utils.SuccessPayload(sensors, "Succesfully retrieved sensors")
+		utils.Response(c, http.StatusOK, result)
+	}
+
+}
+
+func (h *SensorHandler) Update(c *gin.Context) {
+	id := c.Param("id")
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: Which id?

File: app/handlers/sensor.handler.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+		return
+	}
+
+	result = utils.SuccessPayload(nil, "Successfully updated")
+	utils.Response(c, http.StatusOK, result)
+
+}
+
+func (h *SensorHandler) Delete(c *gin.Context) {
+	id := c.Param("id")
```
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: which id?

File: app/models/sensor.model.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
+package models
```
Comment (Tijunel) [inline app/models/sensor.model.go:1]: Nice! Where is lower danger though?

File: app/repositories/sensor/sensor.repository.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+	model "database-ms/app/models"
+	"database-ms/config"
+
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// SensorRepository, used to perform DB operations
+// Interface contains basic operations on sensor document
+// So that, DB operation can be performed easily
```
Comment (Tijunel) [inline app/repositories/sensor/sensor.repository.go]: Find one by Id? Which ID, what is one? Difference between findOne?

 IsSensorAlreadyExists? -> Exists

File: app/handlers/sensor.handler.go (hunk 2 above)
Comment (Tijunel) [inline app/handlers/sensor.handler.go]: We need to create a new sid that is 0-255 but does not exist for any other sensor for the thing. Is this handled?

File: app/repositories/sensor/sensor.repository.go [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+	FindOneById(context.Context, string) (*model.Sensor, error)
+
+	// FindOneByThingIdAndSid
+	FindOneByThingIdAndSid(context.Context, string, int) (*model.Sensor, error)
+
+	// FindByThingIdAndLastUpdate
+	FindByThingIdAndLastUpdate(context.Context, string, int64) ([]*model.Sensor, error)
+
+	// FindOne
+	FindOne(context.Context, interface{}) (*model.Sensor, error)
```
Comment (Tijunel) [inline app/repositories/sensor/sensor.repository.go]: Why is this needed? We will only be fetching all sensors for a thing ID, or a sensor by it's ID.

File: app/repositories/sensor/sensor.repository.go [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+	IsSensorAlreadyExisits(context.Context, bson.ObjectId, int) bool
+}
+
+type SensorRepositoryImp struct {
+	db     *mgo.Session
+	config *config.Configuration
+}
+
+func New(db *mgo.Session, c *config.Configuration) SensorRepository {
```
Comment (Tijunel) [inline app/repositories/sensor/sensor.repository.go]: New what?

File: app/repositories/sensor/sensor.repository.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+	return service.collection().Update(query, change)
+}
+
+func (service *SensorRepositoryImp) Delete(ctx context.Context, id string) error {
+	// TODO Delete from Thing document
+	err := service.collection().RemoveId(bson.ObjectIdHex(id))
+	return err
+}
+
+func (service *SensorRepositoryImp) IsSensorAlreadyExisits(ctx context.Context, thingId bson.ObjectId, sid int) bool {
```
Comment (Tijunel) [inline app/repositories/sensor/sensor.repository.go]: Just use "Exists" for a name.

File: app/services/sensor/sensor.service.go [hunk 17]
```diff
@@ -0,0 +1,86 @@
+package user
+
```
Comment (Tijunel) [inline app/services/sensor/sensor.service.go:2]: This seems quite redundant no? Basically just redefining what was is sensor.repos.go. This could probably be merged with sensor.repos.go. I understand that design and architecture patterns are "best practices", but remember the number one rule - KISS. Also DRY. This is a bunch of extra code that is effectively equivalent to sensor.repos.go. You will even have overhead for having multiple linked calls.

File: controllers/sensor.go [hunk 18]
```diff
@@ -0,0 +1,122 @@
+package controllers
```
Comment (Tijunel) [inline controllers/sensor.go:1]: This seems pretty nice tbh. Straight to the point, it's obvious what's happening. But you are no longer using this? The logic for getting sensor information is obvious and I don't have to dig through multiple files to see what is going on. Having a single file that handles your logic is not a bad thing.

File: routes/api.go [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+	// Routes
+	// TODO: Create middle ware for just token, just key, or both
+	publicEndpoints := c.Group("/database")
+	{
+		// Organization
+		publicEndpoints.GET("/organizations", controllers.GetOrganizations)
+
+		// Sensor
+		// TODO move later as private endpoint
```
Comment (Tijunel) [inline routes/api.go:30]: OK. You created your repo->service->etc model, but what benefits did it really give you? Instead of controllers.CreateSensor, you have sensorAPI.Create. What is worth it? Just some name changing would have solved your problems. I like the idea of separating request types into a separate file, but separating the logic of just getting some sensors into multiple chains of logic seems overkill. I personally think that every line of code added is +1, every line deleted is +2. There are no bonus points for extra code, I give bonus for less code that is clear, concise, and to the point.

File: utils/common.go [hunk 20]
```diff
@@ -0,0 +1,40 @@
+package utils
```
Comment (Tijunel) [inline utils/common.go:1]: Nice, I like having utilities.

File: utils/errors.go [hunk 21]
```diff
@@ -0,0 +1,38 @@
+package utils
+
```
Comment (Tijunel) [inline utils/errors.go:2]: Very nice. Try to be more descriptive though.

File: utils/handler.go [hunk 22]
```diff
@@ -0,0 +1,20 @@
+package utils
```
Comment (Tijunel) [inline utils/handler.go:1]: Nice.

**Review by Tijunel — CHANGES_REQUESTED** [review]
In general, I would stick to the KISS rule. I like the idea of separating out repetitive logic like response types, error types etc. Also, the idea of putting request/response structs is good. I do feel the hierarchical structure of the sensor endpoints is confusing and took me more time than it should to figure out what was going on. Are there really benefits to having "best practices" when it degrades the readability and understandability of the code? My recommendation is to create two files for the sensor stuff: sensor endpoints and sensor helpers. Keep the request logic inside the sensor endpoints file, use the helpers for whatever repetitive logic you find in your endpoints. Easy to understand and does not require digging. Additionally, use long variable names to be descriptive over adding a comment. The best documentation is self-documenting code. Additionally, keep your formatting consistent, there were inconsistent extra lines in functions. I personally try to write my code in chunks of logic, where the comments read as pseudocode. As far as I can tell, this works, but is it the most simple method of making it work? 

Also, we need to create sid for sensors based on what is not currently taken.

File: app/handlers/sensor.handler.go (hunk 7 above)
Comment (justinf34) [reply app/handlers/sensor.handler.go]: 10 is the base and the 64 specifies the integer type that the result must fit into

**Review by Tijunel — APPROVED** [review]
LGTM!

**Review by Tijunel — APPROVED** [review]
